### ISSUE TITLE
Intl lookup can't work without code generation

### ISSUE THREAD

**Issue by jaumard** [dart-lang/intl#409]
Hello,

I wanted to have decentralized traductions instead of having them transformed/generated to dart. But I find myself totally blocked by the intl API and way of doing things. 

Take this example: `String complex(String name, String name2) => Intl.message('$name2 love $name', args: [name, name2], name: 'complex');`

In ARB format it will be generated like this:
```
"complex": "{name2} love {name}",
"@complex": {
  "type": "text",
  "placeholders": {
    "name": {},
    "name2": {}
  }
},
```

To be able to find translations Intl is using `MessageLookup ` under the hood, so I was trying to extends it and make my own lookup based on ARB I have downloaded. 

For that we need to implement this method:
```
String? lookupMessage(String? messageText, String? locale, String? name, List<Object>? args, String? meaning, {MessageIfAbsent? ifAbsent}) {
```

So here I'll receive the sentence `"{name2} love {name}",` and now I'm blocked, because I have a list of `args` but no way to find which one is `name` and which one is `name2`.

I there any workaround or other way I can achieve this ?

**Comment by alan-knight** [comment]
If I understand the question, when you generate ARB from Dart code, it looks at the method signature, so it knows the expected order of the arguments, and they must be passed via args in the same order. When generating from an ARB it also looks at the source code. It uses that information for various purposes - for example, you might have a very large ARB file representing several different projects and want to generate source only for the ones that are used. However, that also tells you the order of the arguments. But you also have the information in the ARB file, in the order of keys in the placeholder map.

In terms of the code generation there are two different sorts of pieces. The messages_all.dart, which contains shared information, and the messages_<locale>.dart, which contains translations for <locale>. Mostly, what messages_all contains is information on how to load and how to interpret the information in the other files, which don't even need to be Dart files at all. And the only part of messages_all that's really generated, in that it's different from run to run, is the list of available translations.

The default version generates the messages_<locale>.dart files as deferred libraries that contain Dart functions for each translations, and messages_all loads those libraries and delegates to them. But there's another format, using JSON. It still generates the same files, but the messages_<locale>.dart files are just long JSON string literals. And the only reason it's done as Dart files rather than data files is that it's a simpler change for projects, If they're data files then there's configuration information about location, loading mechanism, etc.  You should probably take a look at that.

More generally, look at https://github.com/jamesblasco/intl_translation_format , which was a GSOC project by @jamesblasco last year that lets you add arbitrary formats without modifying the intl_translation or intl packages. It doesn't include a pure data format, but it would be easy to add one and to make a messages_all that is or delegates to something that never changes. That could give you some direction, or might be a good framework to use for what you're doing. And I see that he is now a Flutter GDE, so now it's practically his job to help answer questions about it (<grin>, sorry for roping you in, Jaime)

**Comment by jaumard** [comment]
> If I understand the question, when you generate ARB from Dart code, it looks at the method signature, so it knows the expected order of the arguments, and they must be passed via args in the same order. When generating from an ARB it also looks at the source code. It uses that information for various purposes - for example, you might have a very large ARB file representing several different projects and want to generate source only for the ones that are used. However, that also tells you the order of the arguments. But you also have the information in the ARB file, in the order of keys in the placeholder map.

You right except for the placeholder in the ARB, the order of the keys might not match the one of the method (look at my example, they are actually switched, but I can also have multiple time the same key in the translation too).

> But there's another format, using JSON. It still generates the same files, but the messages_.dart files are just long JSON string literals. And the only reason it's done as Dart files rather than data files is that it's a simpler change for projects, If they're data files then there's configuration information about location, loading mechanism, etc. You should probably take a look at that.

But basically what you're telling me is that in order to use Intl without generated code I need to generate code right? :D 

I'm familiar with `intl_translation_format`  and have been in contact with Jaime about it ^^ so I guess the trick is to modify the ARB generation and put the info of the order in the metadata so I can have access to it when reading the ARB file again. 
Might do the trick! 

But shame that Intl couldn't give you back the info in the `lookupMessage` method. I guess my real question (or feature) would be could that be possible in the future ?

**Comment by alan-knight** [comment]
> > If I understand the question, when you generate ARB from Dart code, it looks at the method signature, so it knows the expected order of the arguments, and they must be passed via args in the same order. When generating from an ARB it also looks at the source code. It uses that information for various purposes - for example, you might have a very large ARB file representing several different projects and want to generate source only for the ones that are used. However, that also tells you the order of the arguments. But you also have the information in the ARB file, in the order of keys in the placeholder map.
> 
> You right except for the placeholder in the ARB, the order of the keys might not match the one of the method (look at my example, they are actually switched, but I can also have multiple time the same key in the translation too).

The order in which the placeholders occur in the message body may be switched and there may be multiple occurrences. But the order in which they occur as keys in the placeholders map should match the order they occur in the method signature, and obviously will not have multiple occurrences. They do match in your example. If that's not true in general, it could (and should) be fixed, and with intl_translation_format it wouldn't require a PR to intl_translation. Or, as you say below, you could add extra metadata to specify the placeholder order.

> 
> > But there's another format, using JSON. It still generates the same files, but the messages_.dart files are just long JSON string literals. And the only reason it's done as Dart files rather than data files is that it's a simpler change for projects, If they're data files then there's configuration information about location, loading mechanism, etc. You should probably take a look at that.
> 
> But basically what you're telling me is that in order to use Intl without generated code I need to generate code right? :D

I don't think so, but it depends what you mean. Are you saying you want to extract ARB from the Dart code but then run the program directly? I think the only obstacle to that is the one you mention, that you need to know the translation from positional arguments to named placeholders, which I think is doable, as above. It's not the most efficient, but it could be very convenient for development.

If you want to get the ARB from a different source, e.g. you have translations from a previous version of an app, then at some point you will need to analyze the Dart source to know what the order of the arguments is in Dart. Or you could generate the Dart source from the ARB, which is a whole different sort of code generation.

> 
> I'm familiar with `intl_translation_format` and have been in contact with Jaime about it ^^ so I guess the trick is to modify the ARB generation and put the info of the order in the metadata so I can have access to it when reading the ARB file again.
> Might do the trick!

Great!

> 
> But shame that Intl couldn't give you back the info in the `lookupMessage` method. I guess my real question (or feature) would be could that be possible in the future ?

By 'give you back' do you mean embed it in the ARB or whatever the output format is? That depends on the format. If not, where/when do you expect it?

**Comment by jaumard** [comment]
> The order in which the placeholders occur in the message body may be switched and there may be multiple occurrences. But the order in which they occur as keys in the placeholders map should match the order they occur in the method signature, and obviously will not have multiple occurrences. They do match in your example. If that's not true in general, it could (and should) be fixed, and with intl_translation_format it wouldn't require a PR to intl_translation. Or, as you say below, you could add extra metadata to specify the placeholder order.

Ho yes, the order in the placeholder in metadata is the correct one, but once you parse the JSON you lose that information, unfortunately. 

> I don't think so, but it depends what you mean. Are you saying you want to extract ARB from the Dart code but then run the program directly? I think the only obstacle to that is the one you mention, that you need to know the translation from positional arguments to named placeholders, which I think is doable, as above. It's not the most efficient, but it could be very convenient for development.

Yes, the idea would be to have only the ARB file generated then run the program. 
So yes if I'm the one generating the ARB I will have no issues as I'll add the metadata, if ARB comes from another source then it will not work.

I have added new metadata to give me that order:
```
"@complex": {
    "type": "text",
    "placeholders_order": [
      "name",
      "name2"
    ],
    "placeholders": {
      "name": {},
      "name2": {}
    }
  },
```

With that, it does the job and I'm able to use intl without dart code generated. Meaning I could now download ARB files from a server and have up-to-date translation and I'm not forced to push an update in store just for that. 🦾 😄  

> By 'give you back' do you mean embed it in the ARB or whatever the output format is? That depends on the format. If not, where/when do you expect it?

I have fork `intl_traduction`, migrate it to null safety, and publish it to `intl_generator` because it was blocking me. So I already have updated the ARB generator to put the params order information. So far so good :) 

But what I mean is that to me it would be optimal to have that order directly on the `lookupMessage` method. I see two solutions here: 

- Change the signature of the method to have `Map<String, Object> args` instead of a list, in that case, no need for order anymore, or keep the list but add another one `List<String> argsOrder`. 
- Don't change the signature but change the content of `List<Object> args`, instead of having directly the arguments, having so kind of tuple name|value so we can know to what key corresponds to the arguments.

But I don't know how intl could know the order without generating code to pass it to that function, maybe just ask people to put it, as they do for args.

**Comment by alan-knight** [comment]
> Ho yes, the order in the placeholder in metadata is the correct one, but once you parse the JSON you lose that information, unfortunately.

Are you sure about that? I think parsing the JSON should preserve the key order. I did a quick experiment using jsonDecode in Dartpad and it seems to. I'd be surprised if the VM behaved differently.

> Yes, the idea would be to have only the ARB file generated then run the program.
> So yes if I'm the one generating the ARB I will have no issues as I'll add the metadata, if ARB comes from another source then it will not work.

OK. You also need to control the ARB with the translations. That often comes from a translation tool that might not preserve the data. You could always post-process them along with the original ARB to add it in, but that's an additional step. Only once per translation dump, though.

> But what I mean is that to me it would be optimal to have that order directly on the `lookupMessage` method. I see two solutions here:
> 
> * Change the signature of the method to have `Map<String, Object> args` instead of a list, in that case, no need for order anymore, or keep the list but add another one `List<String> argsOrder`.
> * Don't change the signature but change the content of `List<Object> args`, instead of having directly the arguments, having so kind of tuple name|value so we can know to what key corresponds to the arguments.
> 
> But I don't know how intl could know the order without generating code to pass it to that function, maybe just ask people to put it, as they do for args.

Yes. I wouldn't want to increase the burden on people passing args. Really, I don't want people to have to pass args at all, and back when we had transformers you could avoid that.

**Comment by alan-knight** [comment]
> I have fork `intl_traduction`, migrate it to null safety, and publish it to `intl_generator` because it was blocking me. So I already have updated the ARB generator to put the params order information. So far so good :)

Hey @kevmoo , I saw you got the updated version of intl with stable null-safety pushed. It would be nice to have the same for intl_translation. It seems to be a problem for a lot of people and they're needing to fork.

**Comment by jaumard** [comment]
> Are you sure about that? I think parsing the JSON should preserve the key order. I did a quick experiment using jsonDecode in Dartpad and it seems to. I'd be surprised if the VM behaved differently.

I was sure because I made quick tests about this but now that you said that I double-check and you are right order is conserved... I guess my tests were too quick lol so yeah no need to modify the ARB at all.

> OK. You also need to control the ARB with the translations. That often comes from a translation tool that might not preserve the data. You could always post-process them along with the original ARB to add it in, but that's an additional step. Only once per translation dump, though.

That's is right, and that is my case (I use localazy) and they don't keep metadata, the trick I use is that I always parse metadata on the source ARB file (generally English) where the metadata are normally always present (and if there are not I'll have to trigger an error).

> Yes. I wouldn't want to increase the burden on people passing args. Really, I don't want people to have to pass args at all, and back when we had transformers you could avoid that.

Yes, I agree that it is already a bit painful to put `args` when you have the args already on the calling method. So I can understand that. Just find that having to generate code to do translations is not worth it.  

But based on our conversation I was able to do a POC and everything is working great! 
So far here is what it look like:

For dart only project:
```dart
final delegate = IntlDelegate(
    defaultLocale: 'en',
    dataLoader: (String locale, String flavor) async {
      if (flavor == 'default') {
        return await File('./lib/arb/main_$locale.arb').readAsString();
      }
      return await File('./lib/arb/main_${flavor}_$locale.arb').readAsString();
    },
  );
await delegate.load('fr');
```

For flutter: 
```
TranslationsDelegate<Translations>(
          defaultLocale: 'en',
          currentFlavor: 'flavor1',
          onTranslationsUpdated: () {
            print('TX updated, need rebuild');
          },
          supportedFlavors: ['flavor1'],
          supportedLocales: ['en', 'fr'],
          dataLoader: (locale, flavor) async {
            print('local load $locale and $flavor');
            if (flavor == 'default') {
              return await rootBundle.loadString('assets/arb/translations_$locale.arb');
            }
            return await rootBundle.loadString('assets/arb/translations_${flavor}_$locale.arb');
          },
          updateDataLoader: (locale, flavor) async {
            print('Remote load $locale and $flavor');
            if (locale == 'en' && flavor == 'flavor1') {
              await Future.delayed(Duration(seconds: 10));//simulate some slow network response
              return await rootBundle.loadString('assets/arb/translations_remote_$locale.arb');
            }
            return null;  // no update
          },
          defaultFlavor: 'default',
          translationsBuilder: () => Translations(),
        ),
```

So based on that I'm able to load translations locally based on assets file, and then update them from a remote service to have the last up-to-date version. 

I'll update my `intl_flavors` to use all of that now. Might need to find a better name lol as it will now do much more than just flavor support ^^ any suggestion?

As no update will be done on intl feel free to close this issue if you want.

**Comment by kevmoo** [comment]
CC @natebosch

**Comment by mosuem** [comment]
As far as I see, `intl_flavors` has been discontinued in favor of `jimtl`, correct?

**Comment by jaumard** [comment]
> As far as I see, intl_flavors has been discontinued in favor of jimtl, correct?

`intl_flavors` have been renamed to `jimtl` so that's the same package in the end.
